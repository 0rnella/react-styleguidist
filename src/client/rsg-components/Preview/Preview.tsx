import React, { Component } from 'react';
import ReactDOM from 'react-dom';
import PropTypes from 'prop-types';
import Context from 'rsg-components/Context';
import ReactExample from 'rsg-components/ReactExample';
import PlaygroundError from 'rsg-components/PlaygroundError';

interface PreviewProps {
	code: string;
	filepath: string;
	index: number;
	evalInContext(code: string): () => any;
}

interface PreviewState {
	error: string | null;
	height: number;
}

export default class Preview extends Component<PreviewProps, PreviewState> {
	public static propTypes = {
		code: PropTypes.string.isRequired,
		filepath: PropTypes.string.isRequired,
		evalInContext: PropTypes.func.isRequired,
	};
	public static contextType = Context;

	public state: PreviewState = {
		error: null,
		height: 20,
	};

	private iframeRef = React.createRef<HTMLIFrameElement>();

	public componentDidMount() {
		// Clear console after hot reload, do not clear on the first load
		// to keep any warnings
		if (this.context.codeRevision > 0) {
			// eslint-disable-next-line no-console
			console.clear();
		}

		this.executeCode();
	}

	public shouldComponentUpdate(nextProps: PreviewProps, nextState: PreviewState) {
		return (
			this.state.error !== nextState.error ||
			this.state.height !== nextState.height ||
			this.props.code !== nextProps.code
		);
	}

	public componentDidUpdate(prevProps: PreviewProps) {
		if (this.props.code !== prevProps.code) {
			this.executeCode();
		}
	}

	public componentWillUnmount() {
		this.unmountPreview();
	}

	private getMountNode() {
		// TODO: Do not render to body
		return this.iframeRef.current?.contentWindow?.document?.body;
	}

	private unmountPreview() {
		const node = this.getMountNode();
		if (!node) {
			return;
		}
		ReactDOM.unmountComponentAtNode(node);
	}

	private executeCode() {
		this.setState({
			error: null,
		});

		const { code } = this.props;
		if (!code) {
			return;
		}

		const wrappedComponent: React.FunctionComponentElement<any> = (
			<ReactExample
				code={code}
				evalInContext={this.props.evalInContext}
				onError={this.handleError}
				compilerConfig={this.context.config.compilerConfig}
			/>
		);

		window.requestAnimationFrame(() => {
			const node = this.getMountNode();
			if (!node) {
				return;
			}

			// this.unmountPreview();
			try {
				ReactDOM.render(wrappedComponent, node);
				this.createMutationObserver(node);
				this.handleResize();
			} catch (err) {
				this.handleError(err);
			}
		});
	}

	private createMutationObserver(node: HTMLElement) {
		// TODO: Destroy after unmount
		const observer = new MutationObserver(this.handleResize);
		observer.observe(node, {
			attributes: true,
			attributeOldValue: false,
			characterData: true,
			characterDataOldValue: false,
			childList: true,
			subtree: true,
		});
	}

	// TOOD: Debounce
	private handleResize = () => {
		const node = this.getMountNode();
		if (!node) {
			return;
		}

		this.setState(prevState => {
			const height = node.scrollHeight;
			if (prevState.height === height) {
				return null;
			}
			return { height };
		});
	};

	private handleError = (err: Error) => {
		this.unmountPreview();

		this.setState({
			error: err.toString(),
			height: 0,
		});

		console.error(err); // eslint-disable-line no-console
	};

	public render() {
		const { height, error } = this.state;
		// TODO: Global styles
		// TODO: title
		return (
			<>
				<iframe
					data-testid="mountNode"
					title="TODO"
					ref={this.iframeRef}
					height={height}
					frameBorder={0}
				></iframe>
				{error && <PlaygroundError message={error} />}
			</>
		);
	}
}
