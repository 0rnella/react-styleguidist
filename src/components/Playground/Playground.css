.root {
	composes: border from "../../styles/colors.css";
	overflow: hidden;
	margin-bottom: 1.5rem;
	border-radius: .25rem;
}

.editor {
}

.preview {
	margin-bottom: .2rem;
	padding: 1rem;
}
