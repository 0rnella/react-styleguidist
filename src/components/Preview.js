// Based on https://github.com/joelburget/react-live-editor/blob/master/live-compile.jsx

import React, { PropTypes } from 'react';
import reactTools from 'react-tools';

export default React.createClass({
	displayName: 'Preview',
	propTypes: {
		code: PropTypes.string.isRequired
	},

	getInitialState() {
		return {
			error: null
		};
	},

	componentDidMount: function() {
		this.executeCode();
	},

	componentDidUpdate: function(prevProps) {
		if (this.props.code !== prevProps.code) {
			this.executeCode();
		}
	},

	compileCode: function() {
		// TODO: Babel
		return reactTools.transform(
			// '(function() {' +
				this.props.code,
			// '\n})();'
			{
				harmony: true
			}
		);
	},

	executeCode: function() {
		var mountNode = this.refs.mount.getDOMNode();

		try {
			React.unmountComponentAtNode(mountNode);
		}
		catch (e) {
		}

		this.setState({
			error: null
		});

		try {
			var compiledCode = this.compileCode();
			React.render(eval(compiledCode), mountNode);
		}
		catch (err) {
			React.unmountComponentAtNode(mountNode);
			this.setState({
				error: err.toString()
			});
		}
	},

	renderError() {
		let { error } = this.state;
		if (error) {
			return (
				<div className="playgroundError">{error}</div>
			);
		}
		else {
			return null;
		}
	},

	render: function() {
		return (
			<div>
				<div ref="mount"></div>
				{this.renderError()}
			</div>
		);
	}
});
