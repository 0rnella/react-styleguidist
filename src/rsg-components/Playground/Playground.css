.root {
	composes: border from "../../styles/colors.css";
	overflow: hidden;
	margin-bottom: 30px;
	border-radius: 3px;
}

.editor {
}

.preview {
	margin-bottom: 3px;
	padding: 15px;
}
