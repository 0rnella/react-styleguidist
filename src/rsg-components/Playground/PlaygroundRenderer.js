import React from 'react';
import PropTypes from 'prop-types';
import Styled from 'rsg-components/Styled';

const styles = ({ space, color, borderRadius }) => ({
	root: {
		marginBottom: space[4],
		border: [[1, color.border, 'solid']],
		borderRadius,
	},
	preview: {
		padding: space[2],
	},
	controls: {
		display: 'flex',
		alignItems: 'center',
	},
	toolbar: {
		marginLeft: 'auto',
		marginRight: space[2],
	},
});

export function PlaygroundRenderer({ classes, name, preview, tabButtons, tabBody, toolbar }) {
	return (
		<div className={classes.root}>
			<div className={classes.preview} data-preview={name}>{preview}</div>
			<div className={classes.controls}>
				<div className={classes.tabs}>{tabButtons}</div>
				<div className={classes.toolbar}>{toolbar}</div>
			</div>
			<div className={classes.tab}>{tabBody}</div>
		</div>
	);
}

PlaygroundRenderer.propTypes = {
	classes: PropTypes.object.isRequired,
	name: PropTypes.string.isRequired,
	preview: PropTypes.node.isRequired,
	tabButtons: PropTypes.node.isRequired,
	tabBody: PropTypes.node.isRequired,
	toolbar: PropTypes.node.isRequired,
};

export default Styled(styles)(PlaygroundRenderer);
