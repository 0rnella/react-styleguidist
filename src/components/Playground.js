import React, { PropTypes } from 'react';
import Editor from './Editor';
import Preview from './Preview';

export default React.createClass({
	displayName: 'Playground',
	propTypes: {
		code: PropTypes.string.isRequired,
	},

	getInitialState() {
		return {
			code: this.props.code
		};
	},

	componentWillReceiveProps(nextProps) {
		let { code } = nextProps;
		if (code) {
			this.setState({
				code
			});
		}
	},

	handleChange(newCode) {
		this.setState({
			code: newCode
		});
	},

	render() {
		let { code } = this.state;

		return (
			<div>
				<Editor code={code} onChange={this.handleChange}/>
				<Preview code={code}/>
			</div>
		);
	}
});
