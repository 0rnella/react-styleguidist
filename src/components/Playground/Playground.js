import React, { PropTypes } from 'react';
import Editor from 'components/Editor/Editor';
import Preview from 'components/Preview/Preview';

import s from './Playground.css';

export default React.createClass({
	displayName: 'Playground',
	propTypes: {
		code: PropTypes.string.isRequired,
	},

	getInitialState() {
		return {
			code: this.props.code
		};
	},

	componentWillReceiveProps(nextProps) {
		let { code } = nextProps;
		if (code) {
			this.setState({
				code
			});
		}
	},

	handleChange(newCode) {
		this.setState({
			code: newCode
		});
	},

	render() {
		let { code } = this.state;

		return (
			<div className={s.root}>
				<div className={s.preview}>
					<Preview code={code}/>
				</div>
				<div className={s.editor}>
					<Editor code={code} onChange={this.handleChange}/>
				</div>
			</div>
		);
	}
});
