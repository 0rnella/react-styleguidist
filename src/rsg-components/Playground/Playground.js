import React, { Component } from 'react';
import PropTypes from 'prop-types';
import debounce from 'lodash/debounce';
import Preview from 'rsg-components/Preview';
import Slot from 'rsg-components/Slot';
import PlaygroundRenderer from 'rsg-components/Playground/PlaygroundRenderer';
import { EXAMPLE_TAB_CODE_EDITOR } from '../slots';

export default class Playground extends Component {
	static propTypes = {
		code: PropTypes.string.isRequired,
		evalInContext: PropTypes.func.isRequired,
		index: PropTypes.number.isRequired,
		name: PropTypes.string,
	};
	static contextTypes = {
		config: PropTypes.object.isRequired,
		isolatedExample: PropTypes.bool,
	};

	constructor(props, context) {
		super(props, context);
		const { code } = props;
		const { showCode } = context.config;

		this.handleChange = this.handleChange.bind(this);
		this.handleTabChange = this.handleTabChange.bind(this);

		this.state = {
			code,
			activeTab: showCode ? EXAMPLE_TAB_CODE_EDITOR : undefined,
		};
	}

	componentWillReceiveProps(nextProps) {
		const { code } = nextProps;
		this.setState({
			code,
		});
	}

	shouldComponentUpdate(nextProps, nextState) {
		return nextState.code !== this.state.code || nextState.activeTab !== this.state.activeTab;
	}

	componentWillUnmount() {
		// clear pending changes before unmount
		if (this.queuedChange) {
			this.queuedChange.cancel();
		}
	}

	handleChange(code) {
		// Clear pending changes before proceed
		if (this.queuedChange) {
			this.queuedChange.cancel();
		}

		// Stored update action
		const queuedChange = () =>
			this.setState({
				code,
			});

		const { previewDelay } = this.context.config;

		if (previewDelay) {
			// If previewDelay is enabled debounce the code
			this.queuedChange = debounce(queuedChange, previewDelay);
			this.queuedChange();
		} else {
			// Otherwise execute it
			queuedChange();
		}
	}

	handleTabChange(name) {
		this.setState(state => ({
			activeTab: state.activeTab !== name ? name : undefined,
		}));
	}

	render() {
		const { code, activeTab } = this.state;
		const { evalInContext, index, name } = this.props;
		const { isolatedExample } = this.context;
		return (
			<PlaygroundRenderer
				name={name}
				preview={<Preview code={code} evalInContext={evalInContext} />}
				tabButtons={
					<Slot
						name="exampleTabButtons"
						active={activeTab}
						props={{ onClick: this.handleTabChange }}
					/>
				}
				tabBody={
					<Slot
						name="exampleTabs"
						active={activeTab}
						onlyActive
						props={{ code, onChange: this.handleChange }}
					/>
				}
				toolbar={
					<Slot name="exampleToolbar" props={{ name, isolated: isolatedExample, example: index }} />
				}
			/>
		);
	}
}
