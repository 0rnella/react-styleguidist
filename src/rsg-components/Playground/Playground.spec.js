import test from 'ava';
import React from 'react';
import noop from 'lodash/noop';
import Preview from '../Preview';
import Playground from './Playground';
import PlaygroundRenderer from './PlaygroundRenderer';

const code = '<button>OK</button>';

test('should render component renderer', () => {
	const actual = shallow(
		<Playground
			highlightTheme="base16-light"
			code={code}
			evalInContext={noop}
		/>,
		{
			context: {
				config: {
					showCode: false,
				},
			},
		}
	);

	expect(actual.node, 'to contain',
		<PlaygroundRenderer
			code={code}
			showCode={false}
			highlightTheme="base16-light"
			evalInContext={noop}
		/>
	);
});

test('renderer should render preview', () => {
	const actual = shallow(
		<PlaygroundRenderer
			code={code}
			showCode={false}
			highlightTheme="base16-light"
			evalInContext={noop}
		/>
	);

	expect(actual.node, 'to contain',
		<Preview code={code} />
	);
});
