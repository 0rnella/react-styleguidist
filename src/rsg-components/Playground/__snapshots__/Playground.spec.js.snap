// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`renderer should render preview 1`] = `
<div
  className="root"
>
  <div
    className="previewContainer pizza"
    data-preview="name"
    title="salami"
  >
    <previewContainer>
      <div
        className="preview"
      >
        <div>
          preview
        </div>
      </div>
    </previewContainer>
  </div>
  <div
    className="controls"
  >
    <div>
      <div>
        tab buttons
      </div>
    </div>
    <div
      className="toolbar"
    >
      <div>
        toolbar
      </div>
    </div>
  </div>
  <div>
    <div>
      tab body
    </div>
  </div>
</div>
`;

exports[`should render component renderer 1`] = `
<Styled(Playground)
  code="<button>OK</button>"
  name="name"
  preview={
    <Preview
      code="<button>OK</button>"
      evalInContext={[Function]}
    />
  }
  previewContainer={[Function]}
  previewProps={Object {}}
  tabBody={
    <Slot
      active={undefined}
      name="playgroundTab"
      onlyActive={true}
      props={
        Object {
          "code": "<button>OK</button>",
          "example": 0,
          "isolated": false,
          "name": "name",
          "onChange": [Function],
          "setState": [Function],
          "state": Object {
            "activeTab": undefined,
            "code": "<button>OK</button>",
          },
        }
      }
    />
  }
  tabButtons={
    <Slot
      active={undefined}
      name="playgroundTabButton"
      props={
        Object {
          "code": "<button>OK</button>",
          "example": 0,
          "isolated": false,
          "name": "name",
          "onClick": [Function],
          "setState": [Function],
          "state": Object {
            "activeTab": undefined,
            "code": "<button>OK</button>",
          },
        }
      }
    />
  }
  toolbar={
    <Slot
      name="playgroundToolbarButton"
      props={
        Object {
          "code": "<button>OK</button>",
          "example": 0,
          "isolated": false,
          "name": "name",
          "setState": [Function],
          "state": Object {
            "activeTab": undefined,
            "code": "<button>OK</button>",
          },
        }
      }
    />
  }
/>
`;
